api/types/databaseobject/databaseobject_test.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2024 Gravitational, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package databaseobject
+
+import (
+	"testing"
+
+	"github.com/gravitational/trace"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/defaults"
+	dbobjectv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobject/v1"
+	headerv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/header/v1"
+	"github.com/gravitational/teleport/api/types"
+)
+
+func TestNewDatabaseObject(t *testing.T) {
+	tests := []struct {
+		name          string
+		spec          *dbobjectv1.DatabaseObjectSpec
+		expectedObj   *dbobjectv1.DatabaseObject
+		expectedError error
+	}{
+		{
+			name: "valid object",
+			spec: &dbobjectv1.DatabaseObjectSpec{
+				Protocol:      "postgres",
+				DbServiceName: "test",
+				ObjectKind:    types.KindDatabaseObject,
+				Database:      "test",
+				Schema:        "test",
+				Name:          "test",
+			},
+			expectedError: nil,
+			expectedObj: &dbobjectv1.DatabaseObject{
+				Kind:    types.KindDatabaseObject,
+				Version: types.V1,
+				Metadata: &headerv1.Metadata{
+					Name:      "valid object",
+					Namespace: defaults.Namespace,
+				},
+				Spec: &dbobjectv1.DatabaseObjectSpec{
+					Protocol:      "postgres",
+					DbServiceName: "test",
+					ObjectKind:    types.KindDatabaseObject,
+					Database:      "test",
+					Schema:        "test",
+					Name:          "test",
+				},
+			},
+		},
+		{
+			name:          "missing name",
+			spec:          &dbobjectv1.DatabaseObjectSpec{},
+			expectedError: trace.BadParameter("spec.name: must be non-empty"),
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			obj, err := NewDatabaseObject(tt.name, tt.spec)
+			require.ErrorIs(t, err, tt.expectedError)
```

Comment:
Please avoid comparing complex errors with ErrorIs. These comparisons are brittle and break easily.

Reply:
ErrorAs is much more resilient if you want to check types. trace.IsBadParameter and like methods are similarly resilient.

ErrorContains is less brittle if you want to check that you are hitting a particular error message (say, in cases where you have many BadParameter variants).

Reply:
(Cont, sent too soon)

ErrorsIs is good for guard errors like io.EOF and the like - ie, cases where the entire instance of an error is guaranteed to appear as-is.

Reply:
Agreed about preferred use of different helpers; in this case, I want to check that I'm hitting very particular error condition with my test case, which is why I care both about the error type and message. That makes me believe `require.ErrorIs` is the right choice.